### PR TITLE
Major Data and Form Rework

### PR DESCRIPTION
**Pull request by mtdalton**
Added FormModel to better manage the form data
Added form helper file
Majorly reworked the DynamicForm component to better manage state via the FormModel instance
Updated all the Input fields to use controlled inputs and removed the use of state variables unless absolutely necessary
Removed unused/unnecessary code
Updated Add and Edit screen components to handle loading better
Updated ErrorBlock to better handle errors
Reworked model class
Added BaseModel class (which is the base model class and contains shared functions)
Added DisplayModel and FilterModel classes to handle different model/component cases
Updated FormModel
Removed unused functions from Model class and simplified it to just handle config
Added comments and docblocks to model classes
Updated DataTable, FilterBar and DynamicForm components to use the specific model classes and removed unnecessary functions
Simplified code in Field component
Added Label component (to handle the field name and whether or not that field is required)
Added Label component to all input components
Simplified ErrorBlock component
Added docblocks and comments to list helper
Updated EditScreen to work with the adapted DynamicForm
Added user model, routes and screens
Updated routes and example model file
Fixed state issues and simplified AdminLayout Component
Simplified code in Root file
Fixed error in routes file
Updated authGuard file
Updated the Label component to not display if label is false (which fixes a layout issue in the FilterBar)
Updated models to set fields correctly
Updated route files
Updated package.json file
Updated version number in package.json
Updated build files

### CODE REVIEW

File: src/data/BaseModel.js [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+     * It can get single level values like `data['name']`,
+     * or multi-level values like `data[profile']['first_name']`.
+     *
+     * @param item
+     * @param name
+     * @returns {any}
+     */
+    getValue = (item, name) => {
+        return name.split('.').reduce((o, i) => {
+            return o[i];
```
Comment (BenitoP) [inline src/data/BaseModel.js]: @mtdalton Do you not think we need to have a check here for _undefined_?

File: src/helpers/authGuard.js [hunk 2]
```diff
@@ -21,5 +23,5 @@ export const userDoesNotNeedAuthentication = connectedRouterRedirect({
     // Determine if the user is authenticated or not
     authenticatedSelector: state => !state.auth.accessToken,
     // A nice display name for this check
```
Comment (BenitoP) [inline src/helpers/authGuard.js]: @mtdalton Comment necessary?

File: src/screens/Example/AddExampleScreen.js [hunk 3]
```diff
@@ -12,6 +12,7 @@ const initialProps = {
 const mapDispatchToProps = (dispatch) => ({
     add: (item) => {
         console.log('add');
```
Comment (BenitoP) [inline src/screens/Example/AddExampleScreen.js]: @mtdalton Remove console log

File: src/screens/User/AddUserScreen.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+const initialProps = {
+    title: "Add " + model.type,
+    fields: model.getFormFields(),
+    redirectPath: model.getBaseUrl()
+};
+
+const mapDispatchToProps = (dispatch) => ({
+    add: (item) => {
+        console.log('add');
```
Comment (BenitoP) [inline src/screens/User/AddUserScreen.js]: @mtdalton Remove console log

File: src/screens/Example/AddExampleScreen.js (hunk 3 above)
Comment (mtdalton) [reply src/screens/Example/AddExampleScreen.js]: This console.log needs to stay here. This is on the ExampleScreen that I use to test. Seeing as this add function is not hooked up to an actual API, I log 'add' so I know that it works. This code isn't used and is not accessible when using the RAT in projects.

File: src/screens/User/AddUserScreen.js (hunk 4 above)
Comment (mtdalton) [reply src/screens/User/AddUserScreen.js]: This console.log needs to stay here. This is on the UserScreen that I use to test. Seeing as this add function is not hooked up to an actual API, I log 'add' so I know that it works. This code isn't used and is not accessible when using the RAT in projects.

File: src/data/BaseModel.js (hunk 1 above)
Comment (mtdalton) [reply src/data/BaseModel.js]: We definitely do. I'll add this.

**Pull request merged**